test/e2e/firewall.go

```diff
@@ -0,0 +1,410 @@
+	})
+})
+
+func createOnlyLocalLoadBalancServiceWithPort(jig *ServiceTestJig, ns, serviceName string, port int32, sourceRanges []string) *api.Service {
+	svc := jig.CreateTCPServiceOrFail(ns, func(svc *api.Service) {
+		svc.Spec.Type = api.ServiceTypeLoadBalancer
+		svc.Spec.Ports = []api.ServicePort{{Port: port}}
+		svc.ObjectMeta.Annotations = map[string]string{
+			service.BetaAnnotationExternalTraffic: service.AnnotationValueExternalTrafficLocal}
+		svc.Spec.LoadBalancerSourceRanges = sourceRanges
+	})
+	svc = jig.WaitForLoadBalancerOrFail(ns, serviceName, loadBalancerCreateTimeoutDefault)
+	jig.SanityCheckService(svc, api.ServiceTypeLoadBalancer)
+	return svc
+}
+
+func examineSvcFirewallRules(gceCloud *gcecloud.GCECloud, svc *api.Service) error {
+	nodesTag, err := getNodesTagFromGroupName(framework.TestContext.CloudConfig.NodeInstanceGroup)
+	if err != nil {
+		return err
+	}
+	framework.Logf("nodesTag: %v", nodesTag)
+
+	fw, err := gceCloud.GetFirewall(makeFirewallName(cloudprovider.GetLoadBalancerName(svc)))
+	if err != nil {
+		return err
+	}
+
+	if len(fw.TargetTags) != 1 {
+		return fmt.Errorf("error number of tags, got %v", fw.TargetTags)
+	}
+	if fw.TargetTags[0] != nodesTag {
+		return fmt.Errorf("expect tag: %v, got: %v", nodesTag, fw.TargetTags[0])
+	}
+	framework.Logf("Tags are correct.")
+
+	if err := verifyProtocolsPorts(fw.Allowed, svc.Spec.Ports); err != nil {
+		return err
+	}
+	framework.Logf("Protocols and ports are correct.")
+
+	if err := verifySourceRanges(fw.SourceRanges, svc.Spec.LoadBalancerSourceRanges); err != nil {
+		return err
+	}
+	framework.Logf("Source ranges are correct.")
+
+	return nil
+}
+
+func makeFirewallName(name string) string {
+	return fmt.Sprintf("k8s-fw-%s", name)
+}
+
+func getNodesTagFromGroupName(name string) (string, error) {
+	if !strings.HasSuffix(name, "-group") {
+		return "", fmt.Errorf("node instance group name does not ends with '-group'")
+	}
+	return name[:len(name)-6], nil
+}
+
+func verifyProtocolsPorts(alloweds []*compute.FirewallAllowed, servicePorts []api.ServicePort) error {
+	resultProtocolPorts := []string{}
+	expectedProtocolPorts := []string{}
+
+	for _, allowed := range alloweds {
+		for _, port := range allowed.Ports {
+			resultProtocolPorts = append(resultProtocolPorts, strings.ToLower(allowed.IPProtocol+"/"+port))
+		}
+	}
+	for _, sp := range servicePorts {
+		expectedProtocolPorts = append(expectedProtocolPorts, strings.ToLower(string(sp.Protocol)+"/"+strconv.Itoa(int(sp.Port))))
+	}
+	return sameStringElements(resultProtocolPorts, expectedProtocolPorts)
+}
+
+func verifySourceRanges(resultSourceRanges, expectedSourceRanges []string) error {
+	if expectedSourceRanges == nil {
+		expectedSourceRanges = append(expectedSourceRanges, "0.0.0.0/0")
+	}
+	return sameStringElements(resultSourceRanges, expectedSourceRanges)
+}
+
+func sameStringElements(result, expected []string) error {
+	ex := make(map[string]bool)
```

Comment:
can you just use sets.NewSet to compare? you can basically parse the list into a Set directly, and then use Equals/Diff etc: https://github.com/kubernetes/kubernetes/blob/master/test/e2e/framework/networking_utils.go#L236